core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java

```diff
@@ -0,0 +1,247 @@
+        if (predNode.successors == null) {
+          predNode.successors = new HashSet<>();
+        }
+        predNode.successors.add(nodeID);
+      }
+    }
+
+    if (needsSorting) {
+      // because of the predecessors, we have put the nodeinfos in the map in an incorrect order.
+      // we need to correct that before we try to sort...
+      TreeSet<NodeInfo<ID, N>> tmp = new TreeSet<>(Comparator.comparingInt(a -> a.sourcePosition));
+      tmp.addAll(nodeInfos.values());
+      nodeInfos.clear();
+      tmp.forEach(ni -> nodeInfos.put(ni.id, ni));
+
+      // now we're ready to produce the results
+      sort(nodeInfos, results);
+    } else {
+      // we don't need to sort, but we need to keep the expected behavior of removing the duplicates
+      results = new ArrayList<>(nodeInfos.values());
+    }
+
+    return results.stream().map(ni -> ni.node).collect(Collectors.toList());
+  }
+
+  private void sort(LinkedHashMap<ID, NodeInfo<ID, N>> nodes, List<NodeInfo<ID, N>> results) {
+
+    while (!nodes.isEmpty()) {
+      NodeInfo<ID, N> curr = removeFirstIndependent(nodes);
+      if (curr != null) {
+        // yay, simple. Just add the found independent node to the results.
+        results.add(curr);
+      } else {
+        // ok, there is a cycle in the graph. Let's remove all the nodes in the first cycle we find
+        // from our predecessors map, add them to the result in their original order and try to
+        // continue normally
+
+        // find the first cycle in the predecessors (in the original list order)
+        Iterator<NodeInfo<ID, N>> nexts = nodes.values().iterator();
+        List<NodeInfo<ID, N>> cycle;
+        do {
+          curr = nexts.next();
+          cycle = findCycle(curr, nodes);
+        } while (cycle.isEmpty() && nexts.hasNext());
+
+        // If we ever find a graph that doesn't have any independent node, yet we fail to find a
+        // cycle in it, the universe must be broken.
+        if (cycle.isEmpty()) {
+          throw new IllegalStateException(
+              String.format(
+                  "Failed to find a cycle in a graph that doesn't seem to  have any independent"
+                      + " node. This should never happen. Please file a bug. Current state of the"
+                      + " sorting is: nodes=%s, results=%s",
+                  nodes.toString(), results.toString()));
+        }
+
+        cycle.sort(Comparator.comparingInt(a -> a.sourcePosition));
+
+        for (NodeInfo<ID, N> n : cycle) {
+          removePredecessorMapping(nodes, n);
+          results.add(n);
+        }
+      }
+    }
+  }
+
+  private void removePredecessorMapping(Map<ID, NodeInfo<ID, N>> nodes, NodeInfo<ID, N> node) {
+    if (node.successors != null) {
+      for (ID succ : node.successors) {
+        NodeInfo<ID, N> succNode = nodes.get(succ);
+        if (succNode != null) {
+          succNode.predecessors.remove(node.id);
+        }
+      }
+    }
+    nodes.remove(node.id);
```

Comment:
I can misunderstand but isn't it gets removed in `removeFirstIndependent`? And as far as I know LinkedHashMap throws ConcurrentModificationException if iterator is used after map was changed aside of the iterator.
But if this is intended behavior for different cases then it would be great to leave a comment about that in the code - that remove won't be performed in the iterator case and exceptional case would not occur.

Reply:
Excellent observation! The CME wouldn't occur, because the exact same entry is removed by the iterator in `removeFirstIndependent` prior to calling this method, so the map is never modified by this call. It is just wasteful to call it. I've modified the impl to take this into account.